src/Domain/CoordinateMaps/KerrHorizon.cpp

```diff
@@ -136,11 +146,14 @@ KerrHorizon::inv_jacobian(
 }
 
 template <typename T>
-tt::remove_cvref_wrap_t<T> KerrHorizon::stretch_factor(
+void KerrHorizon::stretch_factor(
+    const gsl::not_null<tt::remove_cvref_wrap_t<T>*> result,
     const std::array<T, 3>& source_coords) const noexcept {
-  const auto coords_mag_sq = dot(source_coords, source_coords);
-  return coords_mag_sq * (coords_mag_sq + spin_mag_sq_) /
-         (coords_mag_sq * coords_mag_sq + square(dot(source_coords, spin_)));
+  tt::remove_cvref_wrap_t<T>& source_coords_sq = *result;
+  source_coords_sq = dot(source_coords, source_coords);
```

Comment:
can you use `dot(make_not_null(&source_coords_sq), ...)` to avoid an allocation here? I'm not sure the allocation in `dot` below (line 156) can be elided

Reply:
this line (153) should be elided without a problem. We have decided against implementing a return by reference `dot` because it made the return type quite complicated. 

for line 156, it is possible that there is an allocation made but I dont see a way around it, we need the value after all.